fw/ss_skb.c

```diff
@@ -1609,3 +1609,289 @@ ss_skb_add_frag(struct sk_buff *skb_head, struct sk_buff *skb, char* addr,
+ * @limit - amount of data we're allowed to send now
+ * @mss - current MSS
+ */
+struct sk_buff *
+ss_skb_hdrm_check(struct sock *sk, struct sk_buff *skb, const unsigned int mss,
+		  const unsigned int limit)
+{
+	struct sk_buff *tmp;
+	struct skb_shared_info *si, *nsi;
+	u8 flags;
+	unsigned int cur_len, proc_len = 0;
+	struct sk_buff *nskb = NULL, *pre_mod_skb = skb->prev;
+
+	T_DBGV("[%d] %s: sk %pK skb %pK mss %u limit %u\n", smp_processor_id(),
+	       __func__, sk, skb, mss, limit);
+
+	skb_queue_walk_from_safe(&sk->sk_write_queue, skb, tmp) {
+		if (proc_len + skb->len > limit)
+			break;
+
+		cur_len = skb->len;
+		if (unlikely(skb_headroom(skb) < SS_SKB_MIN_HDRM)) {
+			si = skb_shinfo(skb);
+
+			nskb = ss_skb_alloc(0);
+			if (!nskb) {
+				goto out_err;
+			}
+			ss_skb_init_for_xmit(nskb);
+
+			/* Tempesta does not utilize @skb->frag_list.
+			 * Incoming SKBs with non-empty frag_list are being
+			 * unrolled in @ss_skb_unroll() */
+			WARN_ON(skb_has_frag_list(skb));
+
+			nsi = skb_shinfo(nskb);
+			if (skb_headlen(skb)) {
+				/* transform linear part of @skb to frag
+				 * at pos @nr_frags of @nskb */
+				__skb_linear_to_frag(skb, nskb, 0,
+						     nsi->nr_frags);
+			}
+
+			if (likely(si->nr_frags < MAX_SKB_FRAGS ||
+				   !skb_headlen(skb))) {
+				/* transfer all @skb frags to @nskb */
+				__skb_frags_transfer(skb, nskb, nsi->nr_frags,
+						     si->nr_frags, 0);
+				__ss_skb_copy_header(skb, nskb);
+
+				__skb_header_release(nskb);
+				__skb_queue_after(&sk->sk_write_queue,
+						  skb, nskb);
+				__skb_unlink(skb, &sk->sk_write_queue);
+
+				ss_sk_trans_mem_account(sk, skb, nskb);
+
+				tcp_skb_tsorted_anchor_cleanup(skb);
+				kfree_skb(skb);
+			} else {
+				/* @skb contains maximum amount of frags already */
+				__ss_skb_copy_header(skb, nskb);
+				TCP_SKB_CB(nskb)->end_seq =
+					TCP_SKB_CB(skb)->seq + skb_headlen(skb);
+
+				flags = TCP_SKB_CB(skb)->tcp_flags;
+				TCP_SKB_CB(nskb)->tcp_flags =
+					flags & ~(TCPHDR_FIN | TCPHDR_PSH);
+
+				ss_sk_trans_mem_account(sk, skb, nskb);
+
+				ss_set_skb_tso_segs(nskb, mss);
+				__skb_header_release(nskb);
+				__skb_queue_after(&sk->sk_write_queue,
+						  skb, nskb);
+
+				nskb = ss_skb_alloc(0);
+				if (!nskb) {
+					tmp = skb->next;
+					__skb_unlink(tmp, &sk->sk_write_queue);
+					sk_wmem_free_skb(sk, tmp);
+					goto out_err;
+				}
+
+				ss_skb_init_for_xmit(nskb);
+				__ss_skb_copy_header(skb, nskb);
+				TCP_SKB_CB(nskb)->seq = TCP_SKB_CB(skb)->seq +
+					skb_headlen(skb);
+				TCP_SKB_CB(nskb)->tcp_flags = flags;
+
+				__skb_frags_transfer(skb, nskb, 0,
+						     si->nr_frags, 0);
+
+				sk_wmem_queued_add(sk, nskb->truesize);
+				sk_mem_charge(sk, nskb->truesize);
+				ss_set_skb_tso_segs(nskb, mss);
+
+				__skb_header_release(nskb);
+				__skb_queue_after(&sk->sk_write_queue,
+						  skb->next, nskb);
+
+				__skb_unlink(skb, &sk->sk_write_queue);
+				tcp_skb_tsorted_anchor_cleanup(skb);
+				kfree_skb(skb);
+			}
```

Comment:
If we need a larger headroom, why not to use something like `__pskb_copy_fclone()` to allocate a new skb with larger headroom and copy `skb` into it? Do we really need all the sophisticated code above?

Reply:
Yes, we can technically use `__pskb_copy_fclone()` or `skb_cow() -> __pskb_expand_head()`, but those functions perform linear data copy, which would be smth like `rep; movsl`. I was trying to avoid explicit copying. Since we would allocate new SKB anyway, i thought moving the linear part as a fragment would be cheaper.